### PR TITLE
Rust code highlighting fixes

### PR DESCRIPTION
**Pull request by HarryLovesCode**
Fixes incorrect highlighting of:
- String escapes
- Feature gates, attributes
- Underscores in numbers

### CODE REVIEW

File: lib/rouge/lexers/rust.rb [hunk 1]
```diff
@@ -78,15 +69,19 @@ def macro_closed?
...
 
       state :root do
-        rule /\n/, Text, :start_line
+        rule /\n/, Text
         mixin :whitespace
         rule /\b(?:#{Rust.keywords.join('|')})\b/, Keyword
         mixin :has_literals
-
+        
+        rule /#!*\[/ do
```
Comment (jneen) [inline lib/rouge/lexers/rust.rb]: should shebangs have to be at the beginning (`\A`) of the file?

**Comment by jneen** [comment]
Hi @HarryLovesCode, could you add some code to `spec/visual/samples/rust` that exercises these fixes and make sure it looks right (by running `bundle exec rackup` and visiting `localhost:9292/rust`)

**Comment by HarryLovesCode** [comment]
Tell me if this needs to be squashed or if there are any concerns you have.

File: lib/rouge/demos/rust [hunk 2]
```diff
@@ -1,12 +1,33 @@
...
+            '+' => accumulator += 1,
+            '-' => accumulator -= 1,
+            '*' => accumulator *= 2,
+            '/' => accumulator /= 2,
+            _ => { /* ignore everything else */ }
         }
     }
+
+    println!("The program \"{}\" calculates the value {}", program, accumulator);
 }
```
Comment (jneen) [inline lib/rouge/demos/rust]: This file is far too long to be a demo now. It is not important to me that the demos compile correctly, just that they show off a little bit of the highlighting. Is there a smaller snippet you could provide here?

**Comment by jneen** [comment]
Hi @HarryLovesCode, I think you added your example to the wrong file - `lib/rouge/demos/rust` is a little taste of what highlighted rust code looks like. I want you to add an example to `spec/visual/samples/rust`, which stress-tests all the little features and corner cases.

**Comment by hugopeixoto** [comment]
Most of the issues addressed by this PR have already been merged:

- [attributes](https://github.com/rouge-ruby/rouge/pull/957)
- [string escapes](https://github.com/rouge-ruby/rouge/pull/1120)
- [underscores **in integers**](https://github.com/rouge-ruby/rouge/pull/984)
- [division operator](https://github.com/rouge-ruby/rouge/pull/1384)

I opened two PRs with the missing features:

- tuple indexing (https://github.com/rouge-ruby/rouge/pull/1580)
- underscores in floats (https://github.com/rouge-ruby/rouge/pull/1581)

**Comment by pyrmont** [comment]
@HarryLovesCode I'm really sorry it took so long to get this all squared away but I've just merged in the two PRs @hugopeixoto opened for the outstanding issues and they'll be part of the next release of Rouge, v3.23.0. That's actually due to come out later today/early tomorrow (depending on your time zone). Given it's been four (!) years, I don't blame you if you've moved on but wanted to keep you in the loop just in case!

@hugopeixoto Thanks so much for both submitting the remaining PRs and doing the yeoman's work of actually tracking down the PRs where the changes had been subsequently added. Really appreciate it!
